### PR TITLE
Add interpretation of <use /> and<defs /> tags

### PR DESCRIPTION
**Pull request by eulertour**
Solves #378.

While writing this I noticed that for rectangles the x and y attributes refer to the center of the Path rather than the top-left corner, so I had to add a workaround for that. I'm open to suggestions if you know of a better way to handle it.

### CODE REVIEW

**Comment by eulertour** [comment]
I tested the PR with this site:
```
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <script src="./build/two.js"></script>
  </head>
  <body>
<svg version='1.1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='53.496596pt' height='27.438854pt' viewBox='144.509481 -15.455973 53.496596 27.438854'>
<defs>
<path id='g3-49' d='M2.929016 -6.37609C2.929016 -6.615193 2.929016 -6.635118 2.699875 -6.635118C2.082192 -5.997509 1.205479 -5.997509 0.886675 -5.997509V-5.688667C1.085928 -5.688667 1.673724 -5.688667 2.191781 -5.947696V-0.787049C2.191781 -0.428394 2.161893 -0.308842 1.265255 -0.308842H0.946451V0C1.295143 -0.029888 2.161893 -0.029888 2.560399 -0.029888S3.825654 -0.029888 4.174346 0V-0.308842H3.855542C2.958904 -0.308842 2.929016 -0.418431 2.929016 -0.787049V-6.37609Z'/>
<path id='g3-54' d='M1.315068 -3.267746V-3.506849C1.315068 -6.027397 2.550436 -6.386052 3.058531 -6.386052C3.297634 -6.386052 3.716065 -6.326276 3.935243 -5.987547C3.785803 -5.987547 3.387298 -5.987547 3.387298 -5.539228C3.387298 -5.230386 3.626401 -5.080946 3.845579 -5.080946C4.004981 -5.080946 4.303861 -5.17061 4.303861 -5.559153C4.303861 -6.156912 3.865504 -6.635118 3.038605 -6.635118C1.763387 -6.635118 0.418431 -5.349938 0.418431 -3.148194C0.418431 -0.488169 1.574097 0.219178 2.500623 0.219178C3.606476 0.219178 4.552927 -0.71731 4.552927 -2.032379C4.552927 -3.297634 3.666252 -4.254047 2.560399 -4.254047C1.882939 -4.254047 1.514321 -3.745953 1.315068 -3.267746ZM2.500623 -0.059776C1.872976 -0.059776 1.574097 -0.657534 1.514321 -0.806974C1.334994 -1.275218 1.334994 -2.072229 1.334994 -2.251557C1.334994 -3.028643 1.653798 -4.024907 2.550436 -4.024907C2.709838 -4.024907 3.16812 -4.024907 3.476961 -3.407223C3.656289 -3.038605 3.656289 -2.530511 3.656289 -2.042341C3.656289 -1.564134 3.656289 -1.066002 3.486924 -0.707347C3.188045 -0.109589 2.729763 -0.059776 2.500623 -0.059776Z'/>
<path id='g3-61' d='M6.844334 -3.257783C6.993773 -3.257783 7.183064 -3.257783 7.183064 -3.457036S6.993773 -3.656289 6.854296 -3.656289H0.886675C0.747198 -3.656289 0.557908 -3.656289 0.557908 -3.457036S0.747198 -3.257783 0.896638 -3.257783H6.844334ZM6.854296 -1.325031C6.993773 -1.325031 7.183064 -1.325031 7.183064 -1.524284S6.993773 -1.723537 6.844334 -1.723537H0.896638C0.747198 -1.723537 0.557908 -1.723537 0.557908 -1.524284S0.747198 -1.325031 0.886675 -1.325031H6.854296Z'/>
<path id='g1-25' d='M2.6401 -3.716065H3.765878C3.437111 -2.241594 3.347447 -1.8132 3.347447 -1.145704C3.347447 -0.996264 3.347447 -0.727273 3.427148 -0.388543C3.526775 0.049813 3.636364 0.109589 3.785803 0.109589C3.985056 0.109589 4.194271 -0.069738 4.194271 -0.268991C4.194271 -0.328767 4.194271 -0.348692 4.134496 -0.488169C3.845579 -1.205479 3.845579 -1.853051 3.845579 -2.132005C3.845579 -2.660025 3.915318 -3.198007 4.024907 -3.716065H5.160648C5.290162 -3.716065 5.648817 -3.716065 5.648817 -4.054795C5.648817 -4.293898 5.439601 -4.293898 5.250311 -4.293898H1.912827C1.693649 -4.293898 1.315068 -4.293898 0.876712 -3.825654C0.52802 -3.437111 0.268991 -2.978829 0.268991 -2.929016C0.268991 -2.919054 0.268991 -2.82939 0.388543 -2.82939C0.468244 -2.82939 0.488169 -2.86924 0.547945 -2.948941C1.036115 -3.716065 1.613948 -3.716065 1.8132 -3.716065H2.381071C2.062267 -2.510585 1.524284 -1.305106 1.105853 -0.398506C1.026152 -0.249066 1.026152 -0.229141 1.026152 -0.159402C1.026152 0.029888 1.185554 0.109589 1.315068 0.109589C1.613948 0.109589 1.693649 -0.169365 1.8132 -0.537983C1.952677 -0.996264 1.952677 -1.016189 2.082192 -1.514321L2.6401 -3.716065Z'/>
<path id='g1-107' d='M2.859278 -6.804483C2.859278 -6.814446 2.859278 -6.914072 2.729763 -6.914072C2.500623 -6.914072 1.77335 -6.834371 1.514321 -6.814446C1.43462 -6.804483 1.325031 -6.794521 1.325031 -6.615193C1.325031 -6.495641 1.414695 -6.495641 1.564134 -6.495641C2.042341 -6.495641 2.062267 -6.425903 2.062267 -6.326276L2.032379 -6.127024L0.587796 -0.388543C0.547945 -0.249066 0.547945 -0.229141 0.547945 -0.169365C0.547945 0.059776 0.747198 0.109589 0.836862 0.109589C0.966376 0.109589 1.115816 0.019925 1.175592 -0.099626C1.225405 -0.18929 1.673724 -2.032379 1.733499 -2.281445C2.072229 -2.251557 2.889166 -2.092154 2.889166 -1.43462C2.889166 -1.364882 2.889166 -1.325031 2.859278 -1.225405C2.839352 -1.105853 2.819427 -0.986301 2.819427 -0.876712C2.819427 -0.288917 3.217933 0.109589 3.73599 0.109589C4.034869 0.109589 4.303861 -0.049813 4.523039 -0.418431C4.772105 -0.856787 4.881694 -1.404732 4.881694 -1.424658C4.881694 -1.524284 4.79203 -1.524284 4.762142 -1.524284C4.662516 -1.524284 4.652553 -1.484433 4.622665 -1.344956C4.423412 -0.617684 4.194271 -0.109589 3.755915 -0.109589C3.566625 -0.109589 3.437111 -0.219178 3.437111 -0.577833C3.437111 -0.747198 3.476961 -0.976339 3.516812 -1.135741C3.556663 -1.305106 3.556663 -1.344956 3.556663 -1.444583C3.556663 -2.092154 2.929016 -2.381071 2.082192 -2.49066C2.391034 -2.669988 2.709838 -2.988792 2.938979 -3.227895C3.417186 -3.755915 3.875467 -4.184309 4.363636 -4.184309C4.423412 -4.184309 4.433375 -4.184309 4.4533 -4.174346C4.572852 -4.154421 4.582814 -4.154421 4.662516 -4.094645C4.682441 -4.084682 4.682441 -4.07472 4.702366 -4.054795C4.224159 -4.024907 4.134496 -3.636364 4.134496 -3.516812C4.134496 -3.35741 4.244085 -3.16812 4.513076 -3.16812C4.772105 -3.16812 5.061021 -3.387298 5.061021 -3.775841C5.061021 -4.07472 4.83188 -4.403487 4.383562 -4.403487C4.104608 -4.403487 3.646326 -4.323786 2.929016 -3.526775C2.590286 -3.148194 2.201743 -2.749689 1.823163 -2.600249L2.859278 -6.804483Z'/>
<path id='g5-88' d='M12.612702 13.947696L13.818182 10.759651H13.569116C13.180573 11.795766 12.124533 12.473225 10.978829 12.772105C10.769614 12.821918 9.793275 13.080946 7.880448 13.080946H1.872976L6.94396 7.13325C7.013699 7.053549 7.033624 7.023661 7.033624 6.973848C7.033624 6.953923 7.033624 6.924035 6.963885 6.824408L2.321295 0.478207H7.780822C9.115816 0.478207 10.022416 0.617684 10.11208 0.637609C10.650062 0.71731 11.516812 0.886675 12.303861 1.384807C12.552927 1.544209 13.230386 1.992528 13.569116 2.799502H13.818182L12.612702 0H0.836862C0.607721 0 0.597758 0.009963 0.56787 0.069738C0.557908 0.099626 0.557908 0.288917 0.557908 0.398506L5.828144 7.611457L0.667497 13.65878C0.56787 13.778331 0.56787 13.828144 0.56787 13.838107C0.56787 13.947696 0.657534 13.947696 0.836862 13.947696H12.612702Z'/>
<path id='g0-49' d='M4.030884 -1.910834C3.66127 -2.343213 3.549689 -2.461768 3.284682 -2.650062C2.859278 -2.956912 2.412951 -3.075467 2.02939 -3.075467C1.164633 -3.075467 0.536986 -2.336239 0.536986 -1.499377C0.536986 -0.683437 1.136737 0.069738 2.008468 0.069738C2.977833 0.069738 3.654296 -0.72528 3.905355 -1.094894C4.274969 -0.662516 4.38655 -0.54396 4.651557 -0.355666C5.076961 -0.048817 5.523288 0.069738 5.906849 0.069738C6.771606 0.069738 7.399253 -0.669489 7.399253 -1.506351C7.399253 -2.322291 6.799502 -3.075467 5.927771 -3.075467C4.958406 -3.075467 4.281943 -2.280448 4.030884 -1.910834ZM4.254047 -1.652802C4.533001 -2.106102 5.139726 -2.810461 5.990535 -2.810461C6.708842 -2.810461 7.203985 -2.175841 7.203985 -1.506351S6.660025 -0.313823 6.018431 -0.313823C5.369863 -0.313823 4.923537 -0.843836 4.254047 -1.652802ZM3.682192 -1.352927C3.403238 -0.899626 2.796513 -0.195268 1.945704 -0.195268C1.227397 -0.195268 0.732254 -0.829888 0.732254 -1.499377S1.276214 -2.691905 1.917808 -2.691905C2.566376 -2.691905 3.012702 -2.161893 3.682192 -1.352927Z'/>
<path id='g2-107' d='M2.182814 -4.630635C2.189788 -4.644583 2.21071 -4.735243 2.21071 -4.742217C2.21071 -4.777086 2.182814 -4.839851 2.099128 -4.839851C1.959651 -4.839851 1.380822 -4.78406 1.206476 -4.770112C1.150685 -4.763138 1.053051 -4.756164 1.053051 -4.609714C1.053051 -4.51208 1.150685 -4.51208 1.234371 -4.51208C1.569116 -4.51208 1.569116 -4.463263 1.569116 -4.407472C1.569116 -4.358655 1.555168 -4.316812 1.54122 -4.254047L0.557908 -0.306849C0.523039 -0.18132 0.523039 -0.167372 0.523039 -0.153425C0.523039 -0.048817 0.606725 0.069738 0.760149 0.069738C0.948443 0.069738 1.039103 -0.069738 1.080946 -0.223163C1.094894 -0.251059 1.39477 -1.478456 1.422665 -1.57609C1.917808 -1.527273 2.315318 -1.366874 2.315318 -1.004234C2.315318 -0.969365 2.315318 -0.934496 2.30137 -0.864757C2.273474 -0.760149 2.273474 -0.72528 2.273474 -0.648568C2.273474 -0.153425 2.677958 0.069738 3.012702 0.069738C3.689166 0.069738 3.898381 -0.990286 3.898381 -0.99726C3.898381 -1.08792 3.807721 -1.08792 3.7868 -1.08792C3.689166 -1.08792 3.682192 -1.053051 3.647323 -0.920548C3.563636 -0.620672 3.375342 -0.125529 3.033624 -0.125529C2.84533 -0.125529 2.789539 -0.299875 2.789539 -0.488169C2.789539 -0.606725 2.789539 -0.620672 2.831382 -0.801993C2.838356 -0.822914 2.866252 -0.941469 2.866252 -1.018182C2.866252 -1.638854 2.02939 -1.736488 1.736488 -1.75741C1.93873 -1.882939 2.196762 -2.113076 2.315318 -2.217684C2.670984 -2.552428 3.019676 -2.880199 3.410212 -2.880199C3.493898 -2.880199 3.584558 -2.859278 3.640349 -2.789539C3.340473 -2.740722 3.277709 -2.503611 3.277709 -2.399004C3.277709 -2.245579 3.396264 -2.140971 3.556663 -2.140971C3.744956 -2.140971 3.954172 -2.294396 3.954172 -2.587298C3.954172 -2.817435 3.7868 -3.075467 3.417186 -3.075467C3.019676 -3.075467 2.657036 -2.789539 2.30137 -2.461768C2.008468 -2.182814 1.778331 -1.966625 1.492403 -1.84807L2.182814 -4.630635Z'/>
<path id='g4-49' d='M2.336239 -4.435367C2.336239 -4.623661 2.322291 -4.630635 2.127024 -4.630635C1.680697 -4.191283 1.046077 -4.184309 0.760149 -4.184309V-3.93325C0.927522 -3.93325 1.387796 -3.93325 1.771357 -4.128518V-0.571856C1.771357 -0.341719 1.771357 -0.251059 1.073973 -0.251059H0.808966V0C0.934496 -0.006974 1.792279 -0.027895 2.050311 -0.027895C2.266501 -0.027895 3.145205 -0.006974 3.29863 0V-0.251059H3.033624C2.336239 -0.251059 2.336239 -0.341719 2.336239 -0.571856V-4.435367Z'/>
<path id='g4-50' d='M3.521793 -1.26924H3.284682C3.263761 -1.115816 3.194022 -0.704359 3.103362 -0.63462C3.047572 -0.592777 2.510585 -0.592777 2.412951 -0.592777H1.129763C1.862017 -1.241345 2.106102 -1.436613 2.524533 -1.764384C3.040598 -2.175841 3.521793 -2.608219 3.521793 -3.270735C3.521793 -4.11457 2.782565 -4.630635 1.889913 -4.630635C1.025156 -4.630635 0.439352 -4.02391 0.439352 -3.382316C0.439352 -3.02665 0.739228 -2.991781 0.808966 -2.991781C0.976339 -2.991781 1.17858 -3.110336 1.17858 -3.361395C1.17858 -3.486924 1.129763 -3.731009 0.767123 -3.731009C0.983313 -4.226152 1.457534 -4.379577 1.785305 -4.379577C2.48269 -4.379577 2.84533 -3.835616 2.84533 -3.270735C2.84533 -2.66401 2.412951 -2.182814 2.189788 -1.931756L0.509091 -0.27198C0.439352 -0.209215 0.439352 -0.195268 0.439352 0H3.312578L3.521793 -1.26924Z'/>
<path id='g4-61' d='M5.355915 -2.343213C5.460523 -2.343213 5.620922 -2.343213 5.620922 -2.517559C5.620922 -2.684932 5.453549 -2.684932 5.362889 -2.684932H0.746202C0.655542 -2.684932 0.488169 -2.684932 0.488169 -2.517559C0.488169 -2.343213 0.648568 -2.343213 0.753176 -2.343213H5.355915ZM5.362889 -0.808966C5.453549 -0.808966 5.620922 -0.808966 5.620922 -0.976339C5.620922 -1.150685 5.460523 -1.150685 5.355915 -1.150685H0.753176C0.648568 -1.150685 0.488169 -1.150685 0.488169 -0.976339C0.488169 -0.808966 0.655542 -0.808966 0.746202 -0.808966H5.362889Z'/>
</defs>
<g id='page1'>
<use x='147.783687' y='-12.453346' xlink:href='#g0-49'/>
<use x='144.559662' y='-9.464569' xlink:href='#g5-88'/>
<use x='144.509481' y='11.982881' xlink:href='#g2-107'/>
<use x='148.913135' y='11.982881' xlink:href='#g4-61'/>
<use x='155.029119' y='11.982881' xlink:href='#g4-49'/>
<use x='164.350391' y='-6.739801' xlink:href='#g3-49'/>
<rect x='161.856265' y='-2.68991' height='0.398484' width='9.969587'/>
<use x='161.856265' y='6.83387' xlink:href='#g1-107'/>
<use x='167.356485' y='3.955777' xlink:href='#g4-50'/>
<use x='175.788699' y='0' xlink:href='#g3-61'/>
<use x='187.500293' y='-6.739801' xlink:href='#g1-25'/>
<use x='193.536709' y='-10.355165' xlink:href='#g4-50'/>
<rect x='187.500293' y='-2.68991' height='0.398484' width='10.505785'/>
<use x='190.262517' y='6.83387' xlink:href='#g3-54'/>
</g>
</svg>
    <div id="two"/>
    <script>
        var elem = document.getElementById('two');
        var two = new Two({
          width: 600,
          height: 600
        }).appendTo(elem);
        shape = two.interpret(document.getElementsByTagName('svg')[0]);
        two.update();
    </script>
  </body>
</html>
```

File: src/utils/interpret-svg.js [hunk 1]
```diff
@@ -303,13 +303,39 @@ var applySvgAttributes = function(node, elem, parentStyles) {
       case 'className':
         elem.classList = value.split(' ');
         break;
+      case 'x':
+      case 'y':
+        elem.translation[key] = parseFloat((value + '').replace('px', ''));
```
Comment (adroitwhiz) [inline src/utils/interpret-svg.js:308]: `parseFloat` silently ignores non-numeric characters, so:
- Stripping `px` from the string is unnecessary
- All values will be interpreted with pixel units, so `2mm` will be treated as 2 pixels, `100%` will be 100 pixels, etc. which may not be desired

File: src/utils/interpret-svg.js [hunk 2]
```diff
@@ -303,13 +303,39 @@ var applySvgAttributes = function(node, elem, parentStyles) {
...
+ */
+var updateDefsCache = function(node, defsCache) {
+  for (var i = 0, l = node.childNodes.length; i < l; i++) {
+    var n = node.childNodes[i];
+    if (!n.id) continue;
+
+    var tag = n.nodeName;
+    if (!tag) continue;
+
+    var tagName = tag.replace(/svg:/ig, '').toLowerCase();
```
Comment (adroitwhiz) [inline src/utils/interpret-svg.js:332]: I think you can use [`localName`](https://developer.mozilla.org/en-US/docs/Web/API/Element/localName) for this

File: src/utils/interpret-svg.js [hunk 3]
```diff
@@ -332,8 +358,22 @@ var getScene = function(node) {
...
+  defsCache: new Map(),
+
   svg: function(node) {
 
+    for (var i = 0, l = node.childNodes.length; i < l; i++) {
+      var n = node.childNodes[i];
+      var tag = n.nodeName;
+      if (!tag) return;
+
+      var tagName = tag.replace(/svg:/ig, '').toLowerCase();
```
Comment (adroitwhiz) [inline src/utils/interpret-svg.js:370]: Same here re: `localName`

File: src/utils/interpret-svg.js [hunk 4]
```diff
@@ -343,15 +383,45 @@ var read = {
...
+        !fullNode.hasAttribute(attr.nodeName)
+      ) {
+        fullNode.setAttribute(attr.nodeName, attr.value);
+      }
+    }
+
+    var tag = fullNode.nodeName;
+    if (!tag) return null;
+
+    var tagName = tag.replace(/svg:/ig, '').toLowerCase();
```
Comment (adroitwhiz) [inline src/utils/interpret-svg.js:423]: Same here re: `localName`

File: src/utils/interpret-svg.js [hunk 5]
```diff
@@ -343,15 +383,45 @@ var read = {
...
+    if (!template) {
+      var error = new TwoError(
+        'unable to find element for reference ' + href + '.'
+      );
+      console.warn(error.name, error.message);
+      return null;
+    }
+    var fullNode = template.cloneNode(true);
+
+    var overwriteAttrs = new Set(
```
Comment (adroitwhiz) [inline src/utils/interpret-svg.js:407]: I'm not sure if `Set` is supported on the range of browsers that Two.js supports (still working this out in #454)

File: src/utils/interpret-svg.js [hunk 6]
```diff
@@ -332,8 +358,22 @@ var getScene = function(node) {
  */
 var read = {
 
+  defsCache: new Map(),
```
Comment (adroitwhiz) [inline src/utils/interpret-svg.js:361]: Not sure about browser compatibility here (re: #454)

File: src/utils/interpret-svg.js (hunk 1 above)
Comment (eulertour) [reply src/utils/interpret-svg.js:308]: Not desired, but this is the way it is for <rect /> tags already.

File: src/utils/interpret-svg.js (hunk 2 above)
Comment (eulertour) [reply src/utils/interpret-svg.js]: Done

File: src/utils/interpret-svg.js (hunk 3 above)
Comment (eulertour) [reply src/utils/interpret-svg.js]: Done

File: src/utils/interpret-svg.js (hunk 4 above)
Comment (eulertour) [reply src/utils/interpret-svg.js]: Done

File: src/utils/interpret-svg.js (hunk 5 above)
Comment (eulertour) [reply src/utils/interpret-svg.js]: I switched to a regular object instead.

File: src/utils/interpret-svg.js (hunk 6 above)
Comment (eulertour) [reply src/utils/interpret-svg.js]: I switched to a regular object.

Comment (jonobr1) [reply src/utils/interpret-svg.js]: Amazing, thanks.

**Comment by jonobr1** [comment]
This looks awesome. Thanks for taking a stab at it! I think everything is good except the `x` and `y` attributes conflicting with `transform`s. I think if both attributes exist (which can be the case for shapes other than the rectangle) there would be a collision. Let me verify this before merging.

**Comment by jonobr1** [comment]
Thanks again for the contribution! This is merged in now.

**Pull request merged**